EventFilter/GEMRawToDigi/plugins/ME0DigiToRawModule.cc

```diff
@@ -0,0 +1,250 @@
+/** \packer for me0
+ *  \author J. Lee - UoS
+ */
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Run.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "DataFormats/FEDRawData/interface/FEDHeader.h"
+#include "DataFormats/FEDRawData/interface/FEDTrailer.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDNumbering.h"
+
+#include "EventFilter/GEMRawToDigi/plugins/ME0DigiToRawModule.h"
+
+using namespace gem;
+
+ME0DigiToRawModule::ME0DigiToRawModule(const edm::ParameterSet & pset):
+  event_type_(pset.getParameter<int>("eventType")),
+  digi_token(consumes<ME0DigiCollection>( pset.getParameter<edm::InputTag>("me0Digi") )),
+  useDBEMap_(pset.getParameter<bool>("useDBEMap"))
+{
+  produces<FEDRawDataCollection>();
+}
+
+void ME0DigiToRawModule::fillDescriptions(edm::ConfigurationDescriptions & descriptions)
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("me0Digi", edm::InputTag("simMuonME0Digis"));
+  desc.add<int>("eventType", 0);
+  desc.add<bool>("useDBEMap", false);
+  descriptions.add("me0Packer", desc);  
+}
+
+std::shared_ptr<ME0ROmap> ME0DigiToRawModule::globalBeginRun(edm::Run const&, edm::EventSetup const& iSetup) const
+{
+  if (useDBEMap_){
+    edm::ESHandle<ME0EMap> me0EMapRcd;
+    iSetup.get<ME0EMapRcd>().get(me0EMapRcd);
+    auto me0EMap = std::make_unique<ME0EMap>(*(me0EMapRcd.product()));
+    return std::make_shared<ME0ROmap>(*(me0EMap->convert()));
```

Comment:
something already partly mentioned in https://github.com/cms-sw/cmssw/pull/21681#issuecomment-357621417, but just to reiterate that making of ME0ROmap better be moved to the event setup.

the implementation here now is apparently doing the following:
- make a local copy of  ME0EMap from ES
- use the local copy to call convert which creates ME0ROmap (and does not delete it)
- make a copy of this created ME0ROmap and handle its lifetime with a shared_ptr.

So, the net cost of the call is in memory for a temporary copy of ME0EMap, a leaked copy of ME0ROmap and in the CPU to call copy constructors.

Also, a somewhat common "repacking" configuration which will call both raw2digi and digi2raw will additionally double the cost compared to what can be available with event setup